backends-velox/src/main/scala/org/apache/spark/sql/catalyst/FlushableHashAggregateRule.scala

```diff
@@ -27,30 +27,69 @@ import org.apache.spark.sql.execution.exchange.ShuffleExchangeLike
 /**
  * To transform regular aggregation to intermediate aggregation that internally enables
  * optimizations such as flushing and abandoning.
- *
- * Currently not in use. Will be enabled via a configuration after necessary verification is done.
  */
 case class FlushableHashAggregateRule(session: SparkSession) extends Rule[SparkPlan] {
   override def apply(plan: SparkPlan): SparkPlan = plan.transformUp {
     case shuffle: ShuffleExchangeLike =>
       // If an exchange follows a hash aggregate in which all functions are in partial mode,
       // then it's safe to convert the hash aggregate to intermediate hash aggregate.
       shuffle.child match {
-        case h: RegularHashAggregateExecTransformer =>
-          if (h.aggregateExpressions.forall(p => p.mode == Partial || p.mode == PartialMerge)) {
-            shuffle.withNewChildren(
-              Seq(FlushableHashAggregateExecTransformer(
-                h.requiredChildDistributionExpressions,
-                h.groupingExpressions,
-                h.aggregateExpressions,
-                h.aggregateAttributes,
-                h.initialInputBufferOffset,
-                h.resultExpressions,
-                h.child
-              )))
-          } else {
-            shuffle
-          }
+        case HashAggPropagatedToShuffle(proj, agg) =>
+          shuffle.withNewChildren(
+            Seq(proj.withNewChildren(Seq(FlushableHashAggregateExecTransformer(
+              agg.requiredChildDistributionExpressions,
+              agg.groupingExpressions,
+              agg.aggregateExpressions,
+              agg.aggregateAttributes,
+              agg.initialInputBufferOffset,
+              agg.resultExpressions,
+              agg.child
+            )))))
+        case HashAggWithShuffle(agg) =>
+          shuffle.withNewChildren(
+            Seq(FlushableHashAggregateExecTransformer(
+              agg.requiredChildDistributionExpressions,
+              agg.groupingExpressions,
+              agg.aggregateExpressions,
+              agg.aggregateAttributes,
+              agg.initialInputBufferOffset,
+              agg.resultExpressions,
+              agg.child
+            )))
+        case _ =>
+          shuffle
       }
   }
 }
+
+object HashAggPropagatedToShuffle {
+  def unapply(
+      plan: SparkPlan): Option[(ProjectExecTransformer, RegularHashAggregateExecTransformer)] = {
+    if (!plan.isInstanceOf[ProjectExecTransformer]) {
+      return None
+    }
+    val proj = plan.asInstanceOf[ProjectExecTransformer]
+    val child = proj.child
+    if (!child.isInstanceOf[RegularHashAggregateExecTransformer]) {
+      return None
+    }
+    val agg = child.asInstanceOf[RegularHashAggregateExecTransformer]
+    if (!agg.aggregateExpressions.forall(p => p.mode == Partial || p.mode == PartialMerge)) {
```

Comment:
Can the correctness of result be ensured with this rule? E.g. if it flushes, the result of distinct count could be wrong?

Reply:
So far the rule would work correctly. The principle is that if a partial aggregation is directly sending data to exchange from mapper side, then a single reducer could always receive partial-aggregated data from several mappers. So enabling flushing on mapper side's agg would act as if adding some mapper instances. 

However there might be some corner cases we should consider, e.g. if partial aggregation is already handing data that is partitioned by the exchange keys, and we may enhance the rule later on for the corner cases. But so far the rule works well on existing tests.